### PR TITLE
[tf] Open http port for json-rpc

### PR DESCRIPTION
**Pull request by sherry-x**
## Motivation

Open port 8080 for json-rpc, create a new load balancer for it. Add listeners on http 80 and 443 for https.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Apply changes to sherryxiao workspace.
new load balancer created: https://us-west-2.console.aws.amazon.com/ec2/v2/home?region=us-west-2#LoadBalancers:search=jsonrpc;sort=loadBalancerName

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/libra/website, and link to your PR here.)

### CODE REVIEW

**Comment by sherry-x** [comment]
Need to open this for fullnode too

**Comment by thefallentree** [comment]
Can we add https listener for this as well?

**Comment by sherry-x** [comment]
@thefallentree HTTPS protocol listener for port 5000? Any reason for that?

**Comment by thefallentree** [comment]
Because json rpc is http based and we should expose https for it. Whether we choose to run port 5000 is an different matter

**Comment by thefallentree** [comment]
And I just checked, https certificate we get from ACM will should on port 5000 too

File: terraform/load-balancing.tf [hunk 1]
```diff
@@ -43,6 +43,37 @@ resource "aws_lb_listener" "validator-ac" {
...
+
+resource "aws_lb_target_group_attachment" "validator-jsonrpc" {
+  count            = var.cluster_test ? 0 : var.num_validators
+  target_group_arn = aws_lb_target_group.validator-jsonrpc.arn
+  target_id        = element(aws_instance.validator.*.id, count.index)
+}
+
+resource "aws_lb_listener" "validator-jsonrpc" {
+  load_balancer_arn = aws_lb.validator-jsonrpc.arn
+  port              = 5000
```
Comment (mgorven) [inline terraform/load-balancing.tf]: Let's use port 80 since it's HTTP.

File: terraform/load-balancing.tf [hunk 2]
```diff
@@ -43,6 +43,37 @@ resource "aws_lb_listener" "validator-ac" {
...
+  name                             = "${terraform.workspace}-jsonrpc"
+  load_balancer_type               = "network"
+  enable_cross_zone_load_balancing = true
+  subnets                          = aws_subnet.testnet.*.id
+}
+
+resource "aws_lb_target_group" "validator-jsonrpc" {
+  name     = "${terraform.workspace}-jsonrpc"
+  protocol = "TCP"
+  port     = 5000
```
Comment (mgorven) [inline terraform/load-balancing.tf]: I'd prefer something like 8080 for the internal port, since it's HTTP.

File: terraform/load-balancing.tf [hunk 3]
```diff
@@ -67,6 +98,30 @@ resource "aws_lb_listener" "fullnode-ac" {
...
+
+resource "aws_lb_target_group_attachment" "fullnode-jsonrpc" {
+  count            = var.num_fullnodes
+  target_group_arn = aws_lb_target_group.fullnode-jsonrpc.arn
+  target_id        = element(aws_instance.fullnode.*.id, count.index)
+}
+
+resource "aws_lb_listener" "fullnode-jsonrpc" {
+  load_balancer_arn = aws_lb.validator-jsonrpc.arn
+  port              = 5001
```
Comment (mgorven) [inline terraform/load-balancing.tf]: Why is the validator and fullnode both being exposed? For AC only the fullnode is.

File: terraform/load-balancing.tf [hunk 4]
```diff
@@ -80,6 +135,19 @@ resource "aws_route53_record" "validator-ac" {
   }
 }
 
+resource "aws_route53_record" "validator-jsonrpc" {
+  count   = local.zone_id == "" ? 0 : 1
+  zone_id = local.zone_id
+  name    = "jsonrpc${local.dns_suffix}"
```
Comment (mgorven) [inline terraform/load-balancing.tf]: I'd do a more generic name, like `client` or `api`.

File: terraform/load-balancing.tf (hunk 3 above)
Comment (sherry-x) [reply terraform/load-balancing.tf]: @mgorven So fullnode will still be using AC? only fullnode to validator is through json-rpc?

File: terraform/load-balancing.tf (hunk 2 above)
Comment (sherry-x) [reply terraform/load-balancing.tf]: This is the number they pick in https://github.com/libra/libra/pull/2861

File: terraform/load-balancing.tf (hunk 3 above)
Comment (phoenix-antigravity) [reply terraform/load-balancing.tf]: no, AC will be removed from both Validator and FN, but enabling json-rpc endpoint on Validator is unnecessary (clients should connect to FNs instead)

File: terraform/load-balancing.tf (hunk 2 above)
Comment (phoenix-antigravity) [reply terraform/load-balancing.tf]: yes, but we're ok to change default port to whatever makes more sense for you guys

**Comment by thefallentree** [comment]
> > Can we add https listener for this as well?
> 
> Let's get this working first, we can add HTTPS later.

There is really no reason not to launch with HTTPS by default. There is going to be an public announcement and everyone will switch to this port (the next release client will already does), I don't want them to use HTTP from day 1, it is just bad practice.

**Comment by sherry-x** [comment]
@thefallentree I can add it

File: terraform/load-balancing.tf [hunk 5]
```diff
@@ -43,6 +43,50 @@ resource "aws_lb_listener" "validator-ac" {
   }
 }
 
+resource "aws_lb" "json-rpc" {
+  name                             = "${terraform.workspace}-jsonrpc"
+  load_balancer_type               = "network"
```
Comment (mgorven) [inline terraform/load-balancing.tf]: This should be able to be an application LB (i.e. L7).

File: terraform/load-balancing.tf [hunk 6]
```diff
@@ -80,6 +125,46 @@ resource "aws_route53_record" "validator-ac" {
   }
 }
 
+resource "aws_route53_record" "json-rpc" {
+  count   = local.zone_id == "" ? 0 : 1
+  zone_id = local.zone_id
+  name    = "client${local.dns_suffix}"
+  type    = "A"
```
Comment (thefallentree) [inline terraform/load-balancing.tf]: can we also add AAAA here?  a lot of fb internal application will use ipv6

File: terraform/load-balancing.tf [hunk 7]
```diff
@@ -43,6 +43,51 @@ resource "aws_lb_listener" "validator-ac" {
   }
 }
 
+resource "aws_lb" "json-rpc" {
+  name                             = "${terraform.workspace}-jsonrpc"
+  load_balancer_type               = "application"
+  enable_cross_zone_load_balancing = true
+  subnets                          = aws_subnet.testnet.*.id
+  security_groups                  = [aws_security_group.validator.id]
+}
```
Comment (thefallentree) [inline terraform/load-balancing.tf]: need to add "dualstack"  here, for AAAA record to work.

Comment (sherry-x) [reply terraform/load-balancing.tf]: @thefallentree sure

Comment (sherry-x) [reply terraform/load-balancing.tf]: will do this in a separate PR

**Pull request closed without merging**